### PR TITLE
Unify dispatch region formation + Conversion to Flow passes

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
There is a weird dependence between the
ConvertToFlowBefore/AfterDispatchRegionFormationPass and the dispatch
region formation itself. Its represents a weird separation of what
cant be moved into dispatches (due to bufferization issue), what needs
to be converted to flow operations and what cannot be converted to
flow operation but still need to be in dispatches.

The only way to resolve these dependencies is to unify these passes.
It also cleans up the dispatch region formation works, allow the
formation to either work on root + fused op or moving individual ops
into the dispatch.

Since the converion to flow happens within dispatch region the
following passes are unnecessary
- `ConvertToFlowAfterDispatchFormation`
- `ConvertToFlowBeforeDisaptchFormation`.
A new pass `ConvertToFlow` is added just for testing the conversion
patterns

### CODE REVIEW

**Comment by benvanik** [comment]
RE insert/extract slice folding: what are the conditions for that happening? One thing we want to make sure of is that we aren't introducing false dependencies - for dense slices it's better for example to produce a slice and then insert it as a separate op as otherwise if a dispatch needs to capture the whole target as an I/O all dispatches touching that target will be serialized. For scattered extracts/slices there's not much we can do, but if there's any chance of being able to take a subspan of a resource we should always do that.

**Comment by MaheshRavishankar** [comment]
> RE insert/extract slice folding: what are the conditions for that happening? One thing we want to make sure of is that we aren't introducing false dependencies - for dense slices it's better for example to produce a slice and then insert it as a separate op as otherwise if a dispatch needs to capture the whole target as an I/O all dispatches touching that target will be serialized. For scattered extracts/slices there's not much we can do, but if there's any chance of being able to take a subspan of a resource we should always do that.

If we want to create the slice and in one op and insert it in another, then thats what will happen before this change, i.e. we dont fuse the `insert_slice` with the op that produces the slice. For "fusion with producer of padding" we need to fuse the `insert_slice` created with its producer. So that the slice created gets written into the right place of the "fused" operation. The latter is not yet prototyped. I can hide it behind a flag for later evaluation and tightening of the controls there. Thanks for the info.

**Comment by benvanik** [comment]
👍 Padding is great to fuse as it's non-contiguous and there's not much anyone else could do with it concurrently anyway. It's more subspan extracts/inserts that we want to ensure we don't fuse from slice/concat ops.

(FYI) Today we can exploit the concurrency when we have that pattern but don't yet place the allocations correctly resulting in an unneeded serialized copy per update. https://github.com/google/iree/issues/7729 is tracking the work - basically, we really want to see those slice + dispatch + update sequences out of flow instead of just dispatches on the full resources (%44/%23/%57). Once that issue is fixed fusing any insertion that could otherwise be updates would always be a pessimization as both are zero-copy but the fusion approach blocks concurrency. It also impacts multi-device as instead of being able to move/stage the smaller subranges of the buffers across devices we'd have to move the entire resource when anything changed.

**Comment by MaheshRavishankar** [comment]
> +1 Padding is great to fuse as it's non-contiguous and there's not much anyone else could do with it concurrently anyway. It's more subspan extracts/inserts that we want to ensure we don't fuse from slice/concat ops.
> 
> (FYI) Today we can exploit the concurrency when we have that pattern but don't yet place the allocations correctly resulting in an unneeded serialized copy per update. #7729 is tracking the work - basically, we really want to see those slice + dispatch + update sequences out of flow instead of just dispatches on the full resources (%44/%23/%57). Once that issue is fixed fusing any insertion that could otherwise be updates would always be a pessimization as both are zero-copy but the fusion approach blocks concurrency. It also impacts multi-device as instead of being able to move/stage the smaller subranges of the buffers across devices we'd have to move the entire resource when anything changed.

Sounds good. I might just drop this. To handle padding we should probably not split the padding early on, but split it during dispatch region formation. Padding needs a fill + insert, and the fill needs to happen separately. Will leave a note for @antiagainst to pick that up when he starts the pad fusion with producer work.

**Comment by benvanik** [comment]
I'm a fan of the cleanup here and if it lets us work around extra padding work then that's cool - just wanted to make sure we were preserving the simple slice/updates :)

**Comment by MaheshRavishankar** [comment]
> I'm a fan of the cleanup here and if it lets us work around extra padding work then that's cool - just wanted to make sure we were preserving the simple slice/updates :)

Oh yeah for sure. Going to land this. I meant Ill drop the part that tries to fuse the `tensor.insert_slice` with its producers. Thats a small part of this.

**Comment by MaheshRavishankar** [comment]
Fun. Regressions here were because the `linalg.fill` wasnt converted to `flow.tensor.splat`. Hopefully this fixes all regressions.

**Review by benvanik — APPROVED** [review]

File: llvm-external-projects/iree-dialects/BUILD [hunk 1]
```diff
@@ -294,7 +294,7 @@ cc_library(
         "@llvm-project//mlir:DialectUtils",
         "@llvm-project//mlir:FuncDialect",
         "@llvm-project//mlir:IR",
-        "@llvm-project//mlir:InferTypeOpIntefaces",
+        "@llvm-project//mlir:InferTypeOpInteface",
```
Comment (okkwon) [inline llvm-external-projects/iree-dialects/BUILD:297]: Inteface -> Interface? I am curious if the typo must be maintained to work properly.

File: iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir [hunk 2]
```diff
@@ -199,3 +199,28 @@ func @elementwise() {
...
+    %2 = flow.dispatch.tensor.load %1, offsets = [%arg0], sizes = [2], strides = [1] : !flow.dispatch.tensor<writeonly:10xf32> -> tensor<2xf32>
+    %7 = linalg.generic {indexing_maps = [#map1, #map1], iterator_types = ["parallel"]} ins(%5 : tensor<2xf32>) outs(%2 : tensor<2xf32>) {
+    ^bb0(%arg1: f32, %arg2: f32):
+      %8 = arith.addf %arg1, %arg1 : f32
+      linalg.yield %8 : f32
+    } -> tensor<2xf32>
+    flow.dispatch.tensor.store %7, %1, offsets = [%arg0], sizes = [2], strides = [1] : tensor<2xf32> -> !flow.dispatch.tensor<writeonly:10xf32>
+  }
+  return
+}
```
Comment (okkwon) [inline iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir:226]: No CHECKs?

File: iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp [hunk 3]
```diff
@@ -319,20 +321,70 @@ struct ConvertTensorFromElementsPattern
...
+      return failure();
+    }
+    SmallVector<SmallVector<Value>> outputShape;
+    ReifyRankedShapedTypeOpInterface reifyShapedTypeInterface =
+        cast<ReifyRankedShapedTypeOpInterface>(reshapeOp.getOperation());
+    if (failed(reifyShapedTypeInterface.reifyResultShapes(rewriter,
+                                                          outputShape))) {
+      return failure();
+    }
+    SmallVector<Value> outputDynamicShapes;
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:342]: Can we use `getDynamicDimValues()'?

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 4]
```diff
@@ -981,8 +981,10 @@ static unsigned decideFusableLinalgOps(FunctionOpInterface funcOp) {
     for (Operation &op : llvm::reverse(block)) {
       // If it is part of a fusion group or root op, ignore it.
       if (hasFusionGroupsAttribute(&op) || hasRootOpAttribute(&op)) continue;
-      // Only look for Linalg ops here.
-      if (!isa<linalg::LinalgOp>(op)) continue;
+      // Only look for Linalg ops here. Avoid moving `linalg.fill` that arent
```
Comment (okkwon) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:984]: nit: arent -> aren't

**Review by okkwon — APPROVED** [review]
Looks good to me. Found some small things.

File: llvm-external-projects/iree-dialects/BUILD (hunk 1 above)
Comment (MaheshRavishankar) [reply llvm-external-projects/iree-dialects/BUILD:297]: Oh thats why CI is failing. Thanks!

**Comment by MaheshRavishankar** [comment]
Have a very interesting case for performance regression here ([link](https://gist.github.com/iree-github-actions-bot/effe3ff220f314e734fa9ba46501d08b)). Most of the above are noise. The only one worth looking into are the MobileBert Fp32 numbers.

For one, the only regress on Vulkan with Adreno. They improve with Mali and on CPU. So I'd consider this not a blocker. The difference though is interesting. Looking at the IR before outlining dispatch region [with](https://gist.github.com/MaheshRavishankar/222a4551c61d1720a2131c5b3e8ce381) and [without](https://gist.github.com/MaheshRavishankar/8de5ef12e6052d4de7c59b3c1ecf2c6e) the changes. I see the only difference is in this chain of ops

```
g1 = gemm1(a, b) // M = 384, N = 128, K = 512
g2 = gemm2(c, d) // M = 384, N = 128, K = 128
result = elementwise_ops(g1, g2, ...)
```

Before this PR, the order was
```
%g2 = gemm2(c, d)
%result = fused_gemm_elementwise(a, b, g2...)
```

After this PR the order is
```
g1 = gemm1(a, b)
%result = fused_gemm_elementwise(c, d, g2...)
```

In theorry both should be exactly identical in terms of performance. The only difference is the K which does not feature in the elementwise operation, so it shouldnt matter which GEMM you fuse this with. The modified code actually should be better cause it does a better job of fusing slices with consumer dispatches so reduces some of that overhead. Wondering if it has something to do with the concurrency..

I am going to check one more case of the MobileSSD that doesnt seem to be noise, but I am not considering this case as a blocker, cause I cant really control this in the fusion, its just using SSA values, and really this both should be equivalent.

FYI @benvanik @antiagainst

File: iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir:226]: Thanks for catching this!

File: iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp:342]: Actually this is resolving the shape of the output in terms of shape of its input. THis is done through the `ReifyRankedShapedTypeInterface` just above this code.

**Comment by MaheshRavishankar** [comment]
The MobileSSD seems like noise. Before and after the change the IR before outlining dispatches is identical.

**Pull request merged**